src/KOKKOS/fix_cmap_kokkos.cpp

```diff
@@ -0,0 +1,925 @@
+  k_crossterm_atom5.modify_device();
+}
+
+/* ----------------------------------------------------------------------
+   initialize one atom's array values, called when atom is created
+------------------------------------------------------------------------- */
+
+template<class DeviceType>
+void FixCMAPKokkos<DeviceType>::set_arrays(int i)
+{
+  k_num_crossterm.sync_host();
+  num_crossterm[i] = 0;
+  k_num_crossterm.modify_host();
+}
+
+/* ----------------------------------------------------------------------
+   pack values in local atom-based arrays for exchange with another proc
+------------------------------------------------------------------------- */
+
+template<class DeviceType>
+int FixCMAPKokkos<DeviceType>::pack_exchange(int i, double *buf)
+{
+  k_num_crossterm.sync_host();
+  k_crossterm_type.sync_host();
+  k_crossterm_atom1.sync_host();
+  k_crossterm_atom2.sync_host();
+  k_crossterm_atom3.sync_host();
+  k_crossterm_atom4.sync_host();
+  k_crossterm_atom5.sync_host();
+
+  int m = FixCMAP::pack_exchange(i,buf);
+
+  k_num_crossterm.modify_host();
+  k_crossterm_type.modify_host();
+  k_crossterm_atom1.modify_host();
+  k_crossterm_atom2.modify_host();
+  k_crossterm_atom3.modify_host();
+  k_crossterm_atom4.modify_host();
+  k_crossterm_atom5.modify_host();
+
+  return m;
+}
+
+/* ----------------------------------------------------------------------
+   unpack values in local atom-based arrays from exchange with another proc
+------------------------------------------------------------------------- */
+
+template<class DeviceType>
+int FixCMAPKokkos<DeviceType>::unpack_exchange(int nlocal, double *buf)
+{
+  k_num_crossterm.sync_host();
+  k_crossterm_type.sync_host();
+  k_crossterm_atom1.sync_host();
+  k_crossterm_atom2.sync_host();
+  k_crossterm_atom3.sync_host();
+  k_crossterm_atom4.sync_host();
+  k_crossterm_atom5.sync_host();
+
+  int m = FixCMAP::unpack_exchange(nlocal,buf);
+
+  k_num_crossterm.modify_host();
+  k_crossterm_type.modify_host();
+  k_crossterm_atom1.modify_host();
+  k_crossterm_atom2.modify_host();
+  k_crossterm_atom3.modify_host();
+  k_crossterm_atom4.modify_host();
+  k_crossterm_atom5.modify_host();
+
+  return m;
+}
+
+/* ----------------------------------------------------------------------
+   pack values in local atom-based array for exchange
+------------------------------------------------------------------------- */
+
+template<class DeviceType>
+int FixCMAPKokkos<DeviceType>::pack_exchange_kokkos(
+   const int &nsend, DAT::tdual_xfloat_2d &k_buf,
+   DAT::tdual_int_1d k_exchange_sendlist, DAT::tdual_int_1d /*k_copylist*/,
+   ExecutionSpace space)
+{
+
+  k_buf.template sync<DeviceType>();
+  //k_copylist.template sync<DeviceType>();
+  k_exchange_sendlist.template sync<DeviceType>();
+
+  auto d_buf = typename ArrayTypes<DeviceType>::t_xfloat_1d_um(
+    k_buf.template view<DeviceType>().data(),
+    k_buf.extent(0)*k_buf.extent(1));
+  //d_copylist = k_copylist.view<DeviceType>();
```

Comment:
@alphataubio I think `copylist` would be needed for pack exchange, are you sure this is correct? For example see https://github.com/lammps/lammps/blob/develop/src/KOKKOS/fix_shake_kokkos.cpp#L1601-L1630.

Reply:
i made the changes for `d_copylist` in latest commit https://github.com/lammps/lammps/pull/4296/commits/4a64b3d9ced759f7c8c189bab6bea3af6036443f, but i dont know how to test `d_copylist` is working properly ?

Reply:
>i dont know how to test d_copylist is working properly ?

If you are running on CPU, you need to set `-pk kokkos comm device` and then test with multiple MPI ranks (exchange isn't called on a single rank).

Reply:
done. tested on openmp and cuda